kernel/mmanutils.c

```diff
@@ -960,15 +960,46 @@ int myst_msync(void* addr, size_t length, int flags)
 
     _rlock(&locked);
     {
+        int prot;
+        bool consistent;
         const size_t n = length / PAGE_SIZE;
-        const uint8_t* page = addr;
+        uint8_t* page = addr;
+        char* prev_pathname = NULL;
 
         for (size_t i = index; i < index + n; i++)
         {
             myst_fdmapping_t* p = &v.fdmappings[i];
-
-            if (p->used == MYST_FDMAPPING_USED)
-                ECHECK(_sync_file(p->fd, p->offset, page, PAGE_SIZE));
+            /* sync only for writable mappings */
+            ECHECK(myst_mman_get_prot(
+                &_mman, page, PAGE_SIZE, &prot, &consistent));
+            if (prot & PROT_WRITE && p->used == MYST_FDMAPPING_USED)
+            {
+                if (p->fd != -1) // valid fd
+                {
+                    // file may have been truncated, ignore failure
+                    _sync_file(p->fd, p->offset, page, PAGE_SIZE);
+                }
+                // invalidated fd, fall back on pathname
+                else
+                {
+                    /* Get file handle the first time, and when the pathname
+                     * field changes*/
+                    if (!prev_pathname ||
+                        !strcmp(prev_pathname, p->pathname->data))
```

Comment:
Does this mean that we only sync on mappings that have the same pathname as the first entry? Just to help me understand, can you explain a bit more on this?

Reply:
The sync happens always, this if block checks whether we need to get a new handle. We fetch a file handle once for the mapping interval which belongs to the same pathname. I have expanded the comment as well.